internal/order/order.go

```diff
@@ -0,0 +1,88 @@
+package order
+
+import "math/rand"
+
+// An Order holds values to return in a randomized round-based sequence such
+// that each value is returned once per round. The sequence is shuffled after
+// each round. The zero value of type Order is an empty Order ready for use.
+type Order[T comparable] struct {
+	a    []T
+	next int
+}
+
+// Next returns the next value in the Order, shuffling first if necessary. It
+// returns the zero value of type T if the Order is empty.
+func (o *Order[T]) Next() T {
+	var t T
+	if len(o.a) == 0 {
+		return t
+	}
+	if o.next == len(o.a) {
+		o.next = 0
+		rand.Shuffle(len(o.a), o.swap)
+	}
+	t = o.a[o.next]
+	o.next++
+	return t
+}
+
+// Add inserts t into a random position in the Order. Depending on where it is
+// inserted, t may or may not be returned in the current round.
+func (o *Order[T]) Add(t T) {
+	o.addAt(t, rand.Intn(len(o.a)+1))
+}
+
+// addAt inserts t at index k, which must be in the range [0, len(o.a)].
+func (o *Order[T]) addAt(t T, k int) {
+	o.a = append(o.a, t)
+	last := len(o.a) - 1
+	if k < o.next {
+		o.swap(o.next, last)
+		o.next++
+	} else {
+		o.swap(k, last)
+	}
+}
+
+// Remove removes the first instance of t from the Order, if any.
+func (o *Order[T]) Remove(t T) {
+	for i := range o.a {
+		if o.a[i] == t {
+			o.removeAt(i)
+			return
+		}
+	}
+}
+
+// removeAt removes the element at index k.
+func (o *Order[T]) removeAt(k int) {
+	if k < o.next {
+		o.next--
+		o.swap(k, o.next)
+		k = o.next
+	}
+	last := len(o.a) - 1
+	o.swap(k, last)
+	o.a = o.a[:last]
+}
+
+// Rand returns a slice of unique elements besides exclude, chosen at random.
+// If there are at least n such elements, Rand returns n of them, or else all.
+func (o *Order[T]) Rand(n int, exclude T) []T {
```

Comment:
I'm not sure Rand is a great name for this. Isn't the whole package supposed to be randomized? What does this do that Next doesn't? Maybe call it like `IndependentSample`? idk

(It's a little weird that this method is here at all, since it feels like it doesn't really have anything to do with the rest of the package, other than that it happens to reuse the underlying slice.)

Reply:
> (It's a little weird that this method is here at all, since it feels like it doesn't really have anything to do with the rest of the package, other than that it happens to reuse the underlying slice.)

It does feel out of place, but the only alternatives I see are:
* to define the order type within the swim package rather than a separate package, so that the fsm can reach into it to perform this task directly (as has been the case so far), or
* to maintain an otherwise unrelated slice of members that can be indexed into within the fsm, which mostly defeats the purpose of abstracting out the ordering functionality in the first place.

The latter doesn't seem productive. The former might be an appropriate middle ground, if having `order` be its own type within the swim package is enough encapsulation.

Reply:
I think having it here in this package is okay. That's why I put that comment in parentheses. I suggest merely acknowledging the awkwardness by giving it a more conspicuous name and maybe also adding a comment along these lines.